posit-dev/positron
# Test parsing and running upgrades

### Intent

* Addresses #2639
* Addresses #2634 
* Improves output for skipped tests (addresses https://github.com/posit-dev/positron/issues/1808#issuecomment-1861241962)
* (No issue for this but actually the most important bit ...) externalizes the tree-sitter queries used to identify test items and load up the test controller.
  - In combination with https://github.com/DavisVaughan/r-tree-sitter, this paves the way for bringing these queries under test. I.e. I hope to test these against various test files in the R package.
  - Also lays the foundation for possible extensibility, i.e. being able to work with `test_that()`-like functions exported by packages other than testthat or internal helpers for a specific package.

### QA Notes

In a workspace that is an R package, click on the flask in the activity bar to activate the test explorer. If the package has testthat tests, the relevant test files will appear in a tree view, for navigation, execution, and display of test results.

I have put up a toy R package that can be used to exercise the bug fixes and features in this PR: https://github.com/jennybc/testfun. I would install this with `pak::pak("jennybc/testfun")`. Another good option is `remotes::install_github("jennybc/testfun")`.

#### More intelligent handling of quotes inside test desc in R test explorer #2639

Run the entire test suite or just the file `test-banana.R` or just the individual test ``'subtraction `still` "works"'``. You can use the ▶️  button in the test explorer tree view or the ▶️  button in the gutter of test file to run a test item (file or test) or the double ▶️  button at the top of the explorer to run all tests.

The test ``'subtraction `still` "works"'`` (and any other tests that have vexing quotes in the description) should run and pass.

#### Provide more support for describe() tests in the test explorer #2634

Use the test explorer to navigate to `test-mathstuff.R` and use the expando to reveal all the tests. Clear all test results from the menu accessed via the `...` in the upper right corner of the TEST tree view. Here is the expected behaviour:

* Running the single `describe("matrix()", { ... })` should work: it should run, there should be green checks for both of its `it()`s and for the `describe()` itself. No _other_ test items in this file should appear as being run; they should still have the grey circle.
* The single `describe("addition()", { ... })` should work _if you run it from the gutter_. A `describe()` with exactly 1 `it()` cannot be run from the explorer's treeview (VS Code tries to run just the child `it()`) and I'm still analyzing whether this is our problem or a bug in VS Code. This is a known issue; see #2805.
* The nested `describe()`s and the enclosing `describe("top-level describe with describe inside", { ... })` cannot be run individually and this is a known issue. For now, such a structure can only be run when the entire file is run. See #2805.

#### Improved output for skipped tests https://github.com/posit-dev/positron/issues/1808#issuecomment-1861241962

Run the single test file `test-apple-wtf.R`. The messages appearing in the TEST OUTPUT tab of the lower panel should appear one to a line and include the test file location where the skip originated.

Good:
```
Reason: I am a skip message: test-apple-wtf.R:2:3
Reason: I am also a skip message: test-apple-wtf.R:9:3
```

Bad:
```
Reason: I am a skip messageReason: I am also a skip message
```
